webapp/src/js/controllers/reports.js

```diff
@@ -336,34 +337,92 @@ angular
       });
     });
 
-    $scope.$on('VerifyReport', function(e, valid) {
-      if (ctrl.selected[0].doc.form) {
-        $scope.setLoadingSubActionBar(true);
+    $scope.$on('VerifyReport', function(e, reportIsValid) {
+      if (!ctrl.selected[0].doc.form) {
+        return;
+      }
+
+      $scope.setLoadingSubActionBar(true);
+
+      const promptUserToConfirmVerification = () => {
+        const verificationTranslationKey = reportIsValid ? 'reports.verify.valid' : 'reports.verify.invalid';
+        return Modal({
+          templateUrl: 'templates/modals/verify_confirm.html',
+          controller: 'VerifyReportModalCtrl',
+          model: {
+            proposedVerificationState: $translate.instant(verificationTranslationKey),
+          },
+        })
+        .then(() => true)
+        .catch(() => false);
+      };
+
+      const shouldReportBeVerified = function (canUserEdit) {
+        // verify if user verifications are allowed
+        if (canUserEdit) {
+          return true;
+        }
+        
+        // don't verify if user can't edit and this is an edit
+        const docHasExistingResult = ctrl.selected[0].doc.verified !== undefined;
+        if (docHasExistingResult) {
+          return false;
+        }
+
+        // verify if this is not an edit and the user accepts  prompt
+        return promptUserToConfirmVerification();
+      };
 
-        if (ctrl.selected[0].doc.contact) {
-          var minifiedContact = lineage.minifyLineage(ctrl.selected[0].doc.contact);
+      const writeVerificationToDoc = function() {
+        const selectedDoc = ctrl.selected[0].doc;
```

Comment:
Redux state is immutable, so by saving this reference and using it throughout, you're effectively using a reference to an old version of the state, then, just a couple of rows lower, the state is updated. 
This causes the UI to be unpredictable, for example this report I just selected as verified incorrect doesn't get the red icon in the corner of the detail page:

![image](https://user-images.githubusercontent.com/35681649/57640483-4f394c00-75bb-11e9-872e-ac702a66c797.png)

Instead, use the reference to the current state `ctrl.selected[0].doc`

Reply:
Okay interesting. Thanks.